packages/@lwc/engine-core/src/framework/api.ts

```diff
@@ -82,12 +82,13 @@ function st(fragment: Element, key: Key): VStatic {
 
 // [fr]agment node
 function fr(key: Key, children: VNodes, stable: 0 | 1): VFragment {
+    const fChildren = children.length > 0 ? [t(''), ...children, t('')] : [];
```

Comment:
When a VFragment node in a set of children VNodes is replaced with `null`, we lose the ability to determine whether the set of children should be marked dynamic when flattening during `allocateChildren`. In the following example, the diffing algo will use the dynamic diffing algo when `$cmp.visible === true`, and the static one when `$cmp.visible === false`.
```javascript
[
    ...,
    $cmp.visible
      ? api_fragment(0, [api_text("Conditional Text")], 0)
      : null,
    ...,
];
```
This leads to issues when the two branches have a different number of VNodes.

One possible solution is to allow for an "empty" fragment as a placeholder. The engine is then able to see that we are in the process of unrendering a VFragment's contents and send the children to the dynamic diffing algo.

See the comment in `rendering.ts` for more context.

Is this appropriate? Can you see any issues with this change?

Reply:
I don't think this is appropriate. The whole point is that you have a vfragment, and you want to mark (delimiting) the position of that vfragment in the dom. The fact that the vfragment is empty doesn't mean much, it might be empty now... but could be non-empty in the next rendering... unless you can guarantee that this is never going to happen.

I'm having a hard time understanding the reasoning to not have those marks.

Reply:
The reasoning not to have those marks is for the use case described in the linked issue https://github.com/salesforce/lwc/issues/3128. When VFragments are used in slots in native shadow, and the content of that vfragment are all assigned to named slots, those extra marks are allocated to the default slot on the component and overrides the default slot content.

The correct behavior in that situation would be to ignore the marks:
```html
<!-- customElement.html -->
<template>
    <slot name="namedSlot">Default named slot content</slot>
    <slot>Default slot content</slot>
</template>

<custom-element>
    #vfragment
        ""
        <div slot="namedSlot">Named Slot Content</div>
        ""
</custom-element>
```
Expected rendered content:
```html
Named Slot Content
Default slot content
```
Actual:
```html
Named Slot Content
"""" <!-- two empty text nodes -->
```

Reply:
> The fact that the vfragment is empty doesn't mean much

With this usage in slots, the fact that the vfragment is empty (unfortunately) does have meaning now; we can't just use `null` like we used to do. It tells us that this _used_ to be a vfragment and we need to take the same path for diffing that we took previously (e.g. if this was/is a stable vfragment, we can do static diffing, if this is not stable, we need to do dynamic diffing).

Reply:
I get that..., `null` is fine IMO.

Reply:
if that's too complicated, I'm ok with this approach, just document it really well in the code.

Reply:
now... looking at this more... I don't think that belongs here... because you don't know here is that fragment is inserted into a slot... it is too generic. Can you do this work somewhere else?

Reply:
Yeah, most of the work is done in `rendering.ts` when we know we're working with a fragment in a slot. That mostly works fine for all but the edge case where the vfragment is replaced with a `null`.

Given the following children: `[div, vfragment(dynamic), div]`, we go down the dynamic diffing path.
If, in the next render, we get `[div, null, div]`, we need something that tells us to continue going down the dynamic diffing path, and `null` tells us nothing.

Having an empty `vfragment(dynamic, [empty])` is one approach that tries to address this from the vfragment side. Another approach might be to add a property on the parent `vnode` or `vm`, but those seemed too generic for dealing with a quirk that's specific to how we've implemented VFragments.